src/Cookie.php

```diff
@@ -0,0 +1,494 @@
+    }
+
+    /**
+     * Checks if there is a value.
+     *
+     * @return bool
+     */
+    public function hasValue()
+    {
+        return isset($this->value);
+    }
+
+    /**
+     * Sets the value.
+     *
+     * @param string|null $value
+     *
+     * @return self
+     */
+    public function withValue($value)
+    {
+        $this->validateValue($value);
+
+        $new = clone $this;
+        $new->value = $value;
+
+        return $new;
+    }
+
+    /**
+     * Returns the max age.
+     *
+     * @return int|null
+     */
+    public function getMaxAge()
+    {
+        return $this->maxAge;
+    }
+
+    /**
+     * Checks if there is a max age.
+     *
+     * @return bool
+     */
+    public function hasMaxAge()
+    {
+        return isset($this->maxAge);
+    }
+
+    /**
+     * Sets the max age.
+     *
+     * @param int|null $maxAge
+     *
+     * @return self
+     */
+    public function withMaxAge($maxAge)
+    {
+        $new = clone $this;
+        $new->maxAge = is_int($maxAge) ? $maxAge : null;
+
+        return $new;
+    }
+
+    /**
+     * Sets both the max age and the expires attributes.
+     *
+     * @param int|\DateTime|null $expiration
+     *
+     * @return self
+     */
+    public function withExpiration($expiration)
+    {
+        $new = clone $this;
+        $new->maxAge = is_int($expiration) ? $expiration : null;
+        $new->expires = $this->normalizeExpires($expiration);
+
+        return $new;
+    }
+
+    /**
+     * Returns the expiration time.
+     *
+     * @return \DateTime|null
+     */
+    public function getExpires()
+    {
+        return $this->expires;
+    }
+
+    /**
+     * Checks if there is an expiration time.
+     *
+     * @return bool
+     */
+    public function hasExpires()
+    {
+        return isset($this->expires);
+    }
+
+    /**
+     * Sets the expires.
+     *
+     * @param \DateTime|null $expires
+     *
+     * @return self
+     */
+    public function withExpires(\DateTime $expires)
+    {
+        $new = clone $this;
+        $new->expires = $expires;
+
+        return $new;
+    }
+
+    /**
+     * Checks if the cookie is expired.
+     *
+     * @return bool
+     */
+    public function isExpired()
+    {
+        return isset($this->expires) and $this->expires < new \DateTime();
+    }
+
+    /**
+     * Returns the domain.
+     *
+     * @return string|null
+     */
+    public function getDomain()
+    {
+        return $this->domain;
+    }
+
+    /**
+     * Checks if there is a domain.
+     *
+     * @return bool
+     */
+    public function hasDomain()
+    {
+        return isset($this->domain);
+    }
+
+    /**
+     * Sets the domain.
+     *
+     * @param string|null $domain
+     *
+     * @return self
+     */
+    public function withDomain($domain)
+    {
+        $new = clone $this;
+        $new->domain = $this->normalizeDomain($domain);
+
+        return $new;
+    }
+
+    /**
+     * Matches a domain.
```

Comment:
this comment says nothing. lets say: "Check whether this cookie is meant for this domain."